kotlinx-coroutines-core/common/src/sync/Semaphore.kt

```diff
@@ -140,74 +142,189 @@ private class SemaphoreImpl(private val permits: Int, acquiredPermits: Int) : Se
     }
 
     /**
-     * This counter indicates a number of available permits if it is non-negative,
-     * or the size with minus sign otherwise. Note, that 32-bit counter is enough here
-     * since the maximal number of available permits is [permits] which is [Int],
-     * and the maximum number of waiting acquirers cannot be greater than 2^31 in any
-     * real application.
+     * This counter indicates the number of available permits if it is positive,
+     * or the negated number of waiters on this semaphore otherwise.
+     * Note, that 32-bit counter is enough here since the maximal number of available
+     * permits is [permits] which is [Int], and the maximum number of waiting acquirers
+     * cannot be greater than 2^31 in any real application.
      */
     private val _availablePermits = atomic(permits - acquiredPermits)
     override val availablePermits: Int get() = max(_availablePermits.value, 0)
 
     private val onCancellationRelease = { _: Throwable -> release() }
 
     override fun tryAcquire(): Boolean {
-        _availablePermits.loop { p ->
+        while (true) {
+            // Get the current number of available permits.
+            val p = _availablePermits.value
+            // Is the number of available permits greater
+            // than the maximal one because of an incorrect
+            // `release()` call without a preceding `acquire()`?
+            // Change it to `permits` and start from the beginning.
+            if (p > permits) {
```

Comment:
Is this check added exclusively for mutex?

If so, it's never executed, because incorrect `unlock` never reaches `release` code-path and gets stuck in owner check

Reply:
No, this is a part of small optimization that leverages Fetch-And-Add in `release`.

Reply:
As mentioned in the comments, this situation may happen due to an incorrect `release()` call, without a preceding `acquire()`.